File: ecolink-back/src/main/java/com/ecolink/model/Activity.java
package com.ecolink.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "activities", indexes = {
    @Index(name = "idx_activity_userId", columnList = "user_id"),
    @Index(name = "idx_activity_createdAt", columnList = "createdAt")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Activity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ActivityType type;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private Integer pointsEarned;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    public enum ActivityType {
        SCHEDULE, COMPLAINT, RECYCLING, ACHIEVEMENT
    }

    @PrePersist
    protected void onCreate() {
        if (this.createdAt == null)   this.createdAt   = LocalDateTime.now();
        if (this.pointsEarned == null) this.pointsEarned = 0;
    }
}

File: flutter_application_1/lib/models/activity.dart
enum ActivityType { schedule, complaint, recycling, achievement }

ActivityType activityTypeFromString(String? value) {
  switch (value) {
    case 'SCHEDULE':
      return ActivityType.schedule;
    case 'COMPLAINT':
      return ActivityType.complaint;
    case 'ACHIEVEMENT':
      return ActivityType.achievement;
    case 'RECYCLING':
    default:
      return ActivityType.recycling;
  }
}

class Activity {
  final int? id;
  final ActivityType type;
  final String description;
  final int pointsEarned;
  final DateTime? createdAt;

  Activity({
    this.id,
    required this.type,
    required this.description,
    this.pointsEarned = 0,
    this.createdAt,
  });

  factory Activity.fromJson(Map<String, dynamic> json) {
    return Activity(
      id: json['id'] as int?,
      type: activityTypeFromString(json['type'] as String?),
      description: json['description'] as String? ?? '',
      pointsEarned: (json['pointsEarned'] as num?)?.toInt() ?? 0,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String)
          : null,
    );
  }
}
